worker/uniter/runner/context/ports.go

```diff
@@ -4,180 +4,178 @@
 }
 
-func validatePortRange(protocol string, fromPort, toPort int) (network.PortRange, error) {
-	// Validate the given range.
-	newRange := network.PortRange{
-		Protocol: strings.ToLower(protocol),
-		FromPort: fromPort,
-		ToPort:   toPort,
-	}
-	if err := newRange.Validate(); err != nil {
-		return network.PortRange{}, err
+func newPortRangeChangeRecorder(unit names.UnitTag, machinePortRanges map[names.UnitTag]map[string][]network.PortRange) *portRangeChangeRecorder {
+	return &portRangeChangeRecorder{
+		machinePortRanges: machinePortRanges,
+		unitTag:           unit,
 	}
-	return newRange, nil
 }
 
-func tryOpenPorts(
-	protocol string,
-	fromPort, toPort int,
-	unitTag names.UnitTag,
-	machinePorts map[network.PortRange]params.RelationUnit,
-	pendingPorts map[PortRange]PortRangeInfo,
-) error {
-	// TODO(dimitern) Once port ranges are linked to relations in
-	// addition to networks, refactor this functions and test it
-	// better to ensure it handles relations properly.
-	relationId := -1
-
-	// Validate the given range.
-	newRange, err := validatePortRange(protocol, fromPort, toPort)
-	if err != nil {
-		return err
-	}
-	rangeKey := PortRange{
-		Ports:      newRange,
-		RelationId: relationId,
+// OpenPortRange registers a request to open the specified port range for the
+// provided endpoint name.
+func (r *portRangeChangeRecorder) OpenPortRange(endpointName string, portRange network.PortRange) error {
+	if err := portRange.Validate(); err != nil {
+		return errors.Trace(err)
 	}
 
-	rangeInfo, isKnown := pendingPorts[rangeKey]
-	if isKnown {
-		if !rangeInfo.ShouldOpen {
-			// If the same range is already pending to be closed, just
-			// mark is pending to be opened.
-			rangeInfo.ShouldOpen = true
-			pendingPorts[rangeKey] = rangeInfo
+	// Ensure port range does not conflict with the ones already recorded
+	// for opening by this unit.
+	if err := r.checkForConflict(endpointName, portRange, r.unitTag, r.pendingOpenRanges, true); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return errors.Annotatef(err, "cannot open %v (unit %q)", portRange, r.unitTag.Id())
 		}
+
+		// Already exists; this is a no-op.
 		return nil
 	}
 
-	// Ensure there are no conflicts with existing ports on the
-	// machine.
-	for portRange, relUnit := range machinePorts {
-		relUnitTag, err := names.ParseUnitTag(relUnit.Unit)
-		if err != nil {
-			return errors.Annotatef(
-				err,
-				"machine ports %v contain invalid unit tag",
-				portRange,
-			)
-		}
-		if newRange.ConflictsWith(portRange) {
-			if portRange == newRange && relUnitTag == unitTag {
-				// The same unit trying to open the same range is just
-				// ignored.
-				return nil
+	// Ensure port range does not conflict with existing open port ranges
+	// for all units deployed to the machine.
+	for otherUnitTag, otherUnitRanges := range r.machinePortRanges {
+		if err := r.checkForConflict(endpointName, portRange, otherUnitTag, otherUnitRanges, false); err != nil {
```

Comment:
Nice extraction.

You *could* even go further and make the checker a type that has these arguments passed to every call, as members.

Reply:
Given the number of changes in the previous refactoring PRs I try to avoid changing too many things in one go ;-)